Add cvmfs-posix-tools

andrey-utkin opened the pull request:
These are tools, modelled after popular file utilities, which offer the convenience of direct effect on CVMFS repo. We believe availability of such commands will make CVMFS administration more attractive to potential adopters.

These tools require "cvmfs_swissknife ingestsql" to be available. ingestsql in turn requires cvmfs_gateway.

One tool is named cvmfs_rsync. There is another script in the git repo called cvmfs_rsync. Please don't confuse them.

Some tests are currently failing which we are looking into, hence draft PR.

DrDaveD commented:
Wow, this is a very large contribution!

I'm not a primary maintainer of cvmfs but have some initial comments.

First, it's not very clear from glancing at the READMEs what these tools add over the standard posix tools.  Maybe that explanation could be added there.  Presumably these are for publishing machines, since cvmfs clients mount read-only, but that's not clear.

Next, there's a name clash with cvmfs_rsync, because the cvmfs-server package already has a /usr/bin/cvmfs_rsync.

Finally, this may be better distributed as a separate package under [cvmfs-contrib](https://cvmfs-contrib.github.io/) instead of integrated in the cvmfs package.

vvolkl commented:
Yes, thanks indeed for preparing this big contribution. Yes, the name clash with cvmfs_rsync is a bit unfortunate, but since cvmfs_rsync does little more than add a few options to rsync, and Jump Trading has been using these tools for a while,  I think we can replace it .

I would certainly package these tool in a separate package, but I have no problem with adding them in this repository.

More comments will follow.

mharvey-jt committed:
Add missing mutexes init, add LoadCatalog

These changes are relevant to ingestsql reliability.

mharvey-jt committed:
Add missing mutexes init, add LoadCatalog

These changes are relevant to ingestsql reliability.

DrDaveD commented:
> cvmfs_rsync does little more than add a few options to rsync, and Jump Trading has been using these tools for a while, I think we can replace it .

But no, there are multiple scripts in place that depend on /usr/bin/cvmfs_rsync to be what it is.  I just had somebody ask me about it within the last week.

andrey-utkin committed:
ingestsql: enable usage without proxy

Before, it refused to work if no proxy was specified (by command line
flag or CVMFS_HTTP_PROXY env var), and it had to be http or https URL.

This patch makes it possible to pass special value DIRECT, or specify
empty string, or leave both flag and env var unset.

andrey-utkin committed:
ingestsql: enable usage without proxy

Before, it refused to work if no proxy was specified (by command line
flag or CVMFS_HTTP_PROXY env var), and it had to be http or https URL.

This patch makes it possible to pass special value DIRECT, or specify
empty string, or leave both flag and env var unset.

DrDaveD commented:
On the other hand, the existing cvmfs_rsync command is part of the cvmfs-server package.  Are these posix tools only expected to be used on client machines?  I would guess so, because otherwise they probably would have noticed that the name was already used.  If that's the case, perhaps it's OK to have a completely different tool by the same name if they're only installed on different machines.  rpm won't let both packages be installed on the same machine.

andrey-utkin committed:
ingestsql: don't mishandle SIGABRT

The signal handler is set for SIGABRT, but it doesn't even exit(1), just
intercepts it and the program execution goes on. It's preferable for the
default handler to be in action, making it exit with distinct non-zero
exit code, possibly dumping core.

andrey-utkin committed:
ingestsql: don't mishandle SIGABRT

The signal handler is set for SIGABRT, but it doesn't even exit(1), just
intercepts it and the program execution goes on. It's preferable for the
default handler to be in action, making it exit with distinct non-zero
exit code, possibly dumping core.

andrey-utkin committed:
ingestsql: turn most assertions into runtime checks

andrey-utkin committed:
ingestsql: turn most assertions into runtime checks

andrey-utkin committed:
ingestsql: when failing, cancel the lease

This avoids a hang for subsequent runs.

andrey-utkin committed:
ingestsql: when failing, cancel the lease

This avoids a hang for subsequent runs.

andrey-utkin commented:
Some fixes coming soon (maybe Friday afternoon).

andrey-utkin committed:
ingestsql: add option -W for wait timeout

andrey-utkin committed:
ingestsql: add option -W for wait timeout

mharvey-jt committed:
Changes relevant to ingestsql reliability

mharvey-jt committed:
Changes relevant to ingestsql reliability

andrey-utkin committed:
Add cvmfs-posix-tools

These are tools which offer the convenience of direct effect on CVMFS
repo.

These tools require "cvmfs_swissknife ingestsql" to be available.
ingestsql manipulates repo state through cvmfs_gateway.

One tool is named cvmfs_rsync. There is another script in the git repo
called cvmfs_rsync. Please don't confuse them.

andrey-utkin committed:
Add cvmfs-posix-tools

These are tools which offer the convenience of direct effect on CVMFS
repo.

These tools require "cvmfs_swissknife ingestsql" to be available.
ingestsql manipulates repo state through cvmfs_gateway.

One tool is named cvmfs_rsync. There is another script in the git repo
called cvmfs_rsync. Please don't confuse them.

andrey-utkin committed:
cvmfs-posix-tools: test: set short TTL

Effective TTL will be 61 second, because s3fanout.cc has harcoded
max-age=61 on .cvmfs* files PUT operations to S3.

Once this will be unhardcoded, this test will run faster.
Still, this test can run successfully on current mainline codebase.

andrey-utkin committed:
cvmfs-posix-tools: test: set short TTL

Effective TTL will be 61 second, because s3fanout.cc has harcoded
max-age=61 on .cvmfs* files PUT operations to S3.

Once this will be unhardcoded, this test will run faster.
Still, this test can run successfully on current mainline codebase.

andrey-utkin committed:
ingestsql: lint

andrey-utkin committed:
catalog_rw: hold lock during Commit(), Transaction()

This chagne is in response to a repeatable SIGSEGV observed while
exercising ingestsql through cvmfs-posix-tools tests.

Observation:

Two threads execute sqlite queries in a race condition, one of them just
"COMMIT". The later one, in libsqlite code, dereferences id->pMethods by
the time the earlier query has zeroed it.

Discussion:

libsqlite mechanisms evidently do not provide thread safety on its own
in the mode used by CVMFS. catalog_rw object already contains a mutex,
so it makes sense to use it in Commit(), Transaction() methods as they
definitely mutate the sqlite context. The immediate issue is that as
part of query execution cleanup, it zeroes in sqlite internal structure
the field id->pMethods, which another query executed by a different
thread is using.

But the calls representing parts of transaction sequence may become
interleaved if multiple threads attempt them at the same time, and the
result will be a runtime error "cannot start a transaction within a
transaction". So higher-level means of synchronisation are also
necessary to completely rectify the observed problem. This is not
addressed here yet.

catalog_mgr_rw FinalizeCatalog() does SyncLock()/SyncUnlock() for a part
of its execution, but leaves catalog->Commit() out. It's not obvious
where the transaction begins. Broadening the locking to all of
FinalizeCatalog() was attempted but results in indefinite hangs.

 Thread 2 received signal SIGSEGV, Segmentation fault.
 [Switching to Thread 5209.5216]
 [Current thread is 2 (Thread 5209.5216)]
 ...
 Completed event: 18900
 Current tick: 7979
 bt
 0x00007dc73a80cb88 in sqlite3OsWrite (id=0x2e1a75e8, pBuf=0x7dc72ebfe2dc, amt=12, offset=41552) at /usr/src/debug/sqlite-3.47.2-2.fc42.x86_64/sqlite3.c:26358
 26358     return id->pMethods->xWrite(id, pBuf, amt, offset);
 (rr) bt
 #0  0x00007dc73a80cb88 in sqlite3OsWrite (id=0x2e1a75e8, pBuf=0x7dc72ebfe2dc, amt=12, offset=41552) at /usr/src/debug/sqlite-3.47.2-2.fc42.x86_64/sqlite3.c:26358
 #1  syncJournal.part.0.lto_priv.0 (pPager=pPager@entry=0x2e1a7370, newHdr=newHdr@entry=0) at /usr/src/debug/sqlite-3.47.2-2.fc42.x86_64/sqlite3.c:61560
 #2  0x00007dc73a80cfaa in syncJournal (pPager=0x2e1a7370, newHdr=0) at /usr/src/debug/sqlite-3.47.2-2.fc42.x86_64/sqlite3.c:63629
 #3  sqlite3PagerCommitPhaseOne.part.0.lto_priv.0 (pPager=0x2e1a7370, zSuper=zSuper@entry=0x0, noSync=noSync@entry=0) at /usr/src/debug/sqlite-3.47.2-2.fc42.x86_64/sqlite3.c:63809
 #4  0x00007dc73a80d17f in sqlite3PagerCommitPhaseOne (pPager=<optimized out>, zSuper=0x0, noSync=0) at /usr/src/debug/sqlite-3.47.2-2.fc42.x86_64/sqlite3.c:63680
 #5  sqlite3BtreeCommitPhaseOne.part.0.lto_priv.0 (p=<optimized out>, zSuperJrnl=0x0) at /usr/src/debug/sqlite-3.47.2-2.fc42.x86_64/sqlite3.c:74958
 #6  0x00007dc73a70d5a0 in sqlite3BtreeCommitPhaseOne (p=<optimized out>, zSuperJrnl=0x0) at /usr/src/debug/sqlite-3.47.2-2.fc42.x86_64/sqlite3.c:74943
 #7  vdbeCommit (db=<optimized out>, p=0x7dc6f4006fb0) at /usr/src/debug/sqlite-3.47.2-2.fc42.x86_64/sqlite3.c:87982
 #8  sqlite3VdbeHalt (p=0x7dc6f4006fb0) at /usr/src/debug/sqlite-3.47.2-2.fc42.x86_64/sqlite3.c:88392
 #9  0x00007dc73a71d985 in sqlite3VdbeExec (p=p@entry=0x7dc6f4006fb0) at /usr/src/debug/sqlite-3.47.2-2.fc42.x86_64/sqlite3.c:94732
 #10 0x00007dc73a71eae4 in sqlite3Step (p=<optimized out>) at /usr/src/debug/sqlite-3.47.2-2.fc42.x86_64/sqlite3.c:91390
 #11 sqlite3_step (pStmt=<optimized out>) at /usr/src/debug/sqlite-3.47.2-2.fc42.x86_64/sqlite3.c:91451
 #12 sqlite3_step (pStmt=<optimized out>) at /usr/src/debug/sqlite-3.47.2-2.fc42.x86_64/sqlite3.c:91440
 #13 0x00000000004e8e71 in sqlite::Sql::Execute (this=0x7dc72ebfe760) at /usr/local/src/cvmfs/cvmfs/sql.cc:43
 #14 0x000000000044595f in catalog::WritableCatalog::UpdateNestedCatalog (this=0x2e1a6a70, path="/dir/inner_dir6", hash=..., size=57344, child_counters=...) at /usr/local/src/cvmfs/cvmfs/catalog_rw.cc:606
 #15 0x000000000042aae9 in catalog::WritableCatalogManager::SingleCatalogUploadCallback (this=0x7ffffb971930, result=...) at /usr/local/src/cvmfs/cvmfs/catalog_mgr_rw.cc:1674
 #16 0x0000000000439812 in BoundCallback<upload::SpoolerResult, catalog::WritableCatalogManager>::operator() (this=0x2e19da00, value=...) at /usr/local/src/cvmfs/cvmfs/util/async.h:91
 #17 0x000000000046ee8a in Observable<upload::SpoolerResult>::NotifyListeners (this=0x2e0a53e0, parameter=...) at /usr/local/src/cvmfs/cvmfs/util/concurrency_impl.h:174
 #18 0x00000000005c3c5b in upload::Spooler::ProcessingCallback (this=0x2e0a53e0, data=...) at /usr/local/src/cvmfs/cvmfs/upload.cc:133
 #19 0x00000000005c4436 in BoundCallback<upload::SpoolerResult, upload::Spooler>::operator() (this=0x2e0b8cb0, value=...) at /usr/local/src/cvmfs/cvmfs/util/async.h:91
 #20 0x000000000046ee8a in Observable<upload::SpoolerResult>::NotifyListeners (this=0x2e0bb5d0, parameter=...) at /usr/local/src/cvmfs/cvmfs/util/concurrency_impl.h:174
 #21 0x000000000046c1eb in IngestionPipeline::OnFileProcessed (this=0x2e0bb5d0, spooler_result=...) at /usr/local/src/cvmfs/cvmfs/ingestion/pipeline.cc:115
 #22 0x00000000004741c4 in BoundCallback<upload::SpoolerResult, IngestionPipeline>::operator() (this=0x2e0bc8c0, value=...) at /usr/local/src/cvmfs/cvmfs/util/async.h:91
 #23 0x000000000046ee8a in Observable<upload::SpoolerResult>::NotifyListeners (this=0x2e0badd0, parameter=...) at /usr/local/src/cvmfs/cvmfs/util/concurrency_impl.h:174
 #24 0x0000000000478c13 in TaskRegister::Process (this=0x2e0badc0, file_item=0x2e1cc850) at /usr/local/src/cvmfs/cvmfs/ingestion/task_register.cc:30
 #25 0x0000000000470c5d in TubeConsumer<FileItem>::MainConsumer (data=0x2e0badc0) at /usr/local/src/cvmfs/cvmfs/ingestion/task.h:56
 #26 0x00007dc739c23f54 in start_thread (arg=<optimized out>) at pthread_create.c:448
 #27 0x00007dc739ca7154 in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:100
 (rr)

 (rr) print $_siginfo
 $1 = {
   si_signo = 11,
   si_errno = 0,
   si_code = 1,
   _sifields = {
     ...
     _sigfault = {
       si_addr = 0x18
     },
   }
 }

 (rr) watch -l (void*) id->pMethods
 (rr) rc
 Continuing.
 [Switching to Thread 5209.5209]
 [Current thread is 1 (Thread 5209.5209)]
 #0  0x00007dc73a6e0af3 in memset (__dest=0x2e1a75e8, __ch=0, __len=120) at /usr/include/bits/string_fortified.h:59
 #1  closeUnixFile (id=0x2e1a75e8) at /usr/src/debug/sqlite-3.47.2-2.fc42.x86_64/sqlite3.c:40656
 #2  0x00007dc73a6e0cad in nolockClose (id=<optimized out>) at /usr/src/debug/sqlite-3.47.2-2.fc42.x86_64/sqlite3.c:40733
 Completed event: 18833
 Current tick: 6377324

 Thread 1 hit Hardware watchpoint 3: -location (void*) id->pMethods

 Old value = (void *) 0x0
 New value = (void *) 0x7dc73a8497c0 <nolockIoMethods>
 0x00007dc73a6e0af3 in memset (__dest=0x2e1a75e8, __ch=0, __len=120) at /usr/include/bits/string_fortified.h:59
 warning: Source file is more recent than executable.
 59        return __builtin___memset_chk (__dest, __ch, __len,
 (rr) bt
 #0  0x00007dc73a6e0af3 in memset (__dest=0x2e1a75e8, __ch=0, __len=120) at /usr/include/bits/string_fortified.h:59
 #1  closeUnixFile (id=0x2e1a75e8) at /usr/src/debug/sqlite-3.47.2-2.fc42.x86_64/sqlite3.c:40656
 #2  0x00007dc73a6e0cad in nolockClose (id=<optimized out>) at /usr/src/debug/sqlite-3.47.2-2.fc42.x86_64/sqlite3.c:40733
 #3  0x00007dc73a6eee5b in sqlite3OsClose (pId=0x2e1a75e8) at /usr/src/debug/sqlite-3.47.2-2.fc42.x86_64/sqlite3.c:26348
 #4  pager_end_transaction (pPager=0x2e1a7370, hasSuper=<optimized out>, bCommit=1) at /usr/src/debug/sqlite-3.47.2-2.fc42.x86_64/sqlite3.c:59323
 #5  0x00007dc73a6fc0a2 in sqlite3PagerCommitPhaseTwo (pPager=0x2e1a7370) at /usr/src/debug/sqlite-3.47.2-2.fc42.x86_64/sqlite3.c:63935
 #6  sqlite3BtreeCommitPhaseTwo (p=0x2e1a1ac0, bCleanup=<optimized out>) at /usr/src/debug/sqlite-3.47.2-2.fc42.x86_64/sqlite3.c:75044
 #7  0x00007dc73a70d5d3 in vdbeCommit (db=<optimized out>, p=0x2e1b2100) at /usr/src/debug/sqlite-3.47.2-2.fc42.x86_64/sqlite3.c:87994
 #8  sqlite3VdbeHalt (p=0x2e1b2100) at /usr/src/debug/sqlite-3.47.2-2.fc42.x86_64/sqlite3.c:88392
 #9  0x00007dc73a71af85 in sqlite3VdbeExec (p=p@entry=0x2e1b2100) at /usr/src/debug/sqlite-3.47.2-2.fc42.x86_64/sqlite3.c:97385
 #10 0x00007dc73a71eae4 in sqlite3Step (p=<optimized out>) at /usr/src/debug/sqlite-3.47.2-2.fc42.x86_64/sqlite3.c:91390
 #11 sqlite3_step (pStmt=<optimized out>) at /usr/src/debug/sqlite-3.47.2-2.fc42.x86_64/sqlite3.c:91451
 #12 sqlite3_step (pStmt=<optimized out>) at /usr/src/debug/sqlite-3.47.2-2.fc42.x86_64/sqlite3.c:91440
 #13 0x00000000004e8e71 in sqlite::Sql::Execute (this=0x2e1b5ef0) at /usr/local/src/cvmfs/cvmfs/sql.cc:43
 #14 0x00000000004418ac in sqlite::Database<catalog::CatalogDatabase>::CommitTransaction (this=0x2e19b2b0) at /usr/local/src/cvmfs/cvmfs/sql_impl.h:275
 #15 0x00000000004433d0 in catalog::WritableCatalog::Commit (this=0x2e1a6a70) at /usr/local/src/cvmfs/cvmfs/catalog_rw.cc:83
 #16 0x0000000000428e8e in catalog::WritableCatalogManager::FinalizeCatalog (this=0x7ffffb971930, catalog=0x2e1a6a70, stop_for_tweaks=false) at /usr/local/src/cvmfs/cvmfs/catalog_mgr_rw.cc:1284
 #17 0x000000000042a6f4 in catalog::WritableCatalogManager::ScheduleReadyCatalogs (this=0x7ffffb971930) at /usr/local/src/cvmfs/cvmfs/catalog_mgr_rw.cc:1632
 #18 0x000000000052eeb0 in swissknife::IngestSQL::do_additions (this=0x2e09f7a0, all_dirs=std::map with 12 elements = {...}, all_files=std::map with 9 elements = {...}, all_symlinks=std::map with 0 elements,
     lease_path="", catalog_manager=...) at /usr/local/src/cvmfs/cvmfs/swissknife_ingestsql.cc:1133
 #19 0x000000000052e6ad in swissknife::IngestSQL::process_sqlite (this=0x2e09f7a0, dbs=std::vector of length 1, capacity 1 = {...}, catalog_manager=..., allow_additions=true, allow_deletions=true, lease_path="",
     additional_prefix="") at /usr/local/src/cvmfs/cvmfs/swissknife_ingestsql.cc:1038
 #20 0x000000000052db9a in swissknife::IngestSQL::Main (this=0x2e09f7a0, args=std::map with 12 elements = {...}) at /usr/local/src/cvmfs/cvmfs/swissknife_ingestsql.cc:889
 #21 0x00000000005540e6 in main (argc=24, argv=0x7ffffb9731b8) at /usr/local/src/cvmfs/cvmfs/swissknife_main.cc:190

andrey-utkin committed:
catalog_rw: hold lock during Commit(), Transaction()

This chagne is in response to a repeatable SIGSEGV observed while
exercising ingestsql through cvmfs-posix-tools tests.

Observation:

Two threads execute sqlite queries in a race condition, one of them just
"COMMIT". The later one, in libsqlite code, dereferences id->pMethods by
the time the earlier query has zeroed it.

Discussion:

libsqlite mechanisms evidently do not provide thread safety on its own
in the mode used by CVMFS. catalog_rw object already contains a mutex,
so it makes sense to use it in Commit(), Transaction() methods as they
definitely mutate the sqlite context. The immediate issue is that as
part of query execution cleanup, it zeroes in sqlite internal structure
the field id->pMethods, which another query executed by a different
thread is using.

But the calls representing parts of transaction sequence may become
interleaved if multiple threads attempt them at the same time, and the
result will be a runtime error "cannot start a transaction within a
transaction". So higher-level means of synchronisation are also
necessary to completely rectify the observed problem. This is not
addressed here yet.

catalog_mgr_rw FinalizeCatalog() does SyncLock()/SyncUnlock() for a part
of its execution, but leaves catalog->Commit() out. It's not obvious
where the transaction begins. Broadening the locking to all of
FinalizeCatalog() was attempted but results in indefinite hangs.

 Thread 2 received signal SIGSEGV, Segmentation fault.
 [Switching to Thread 5209.5216]
 [Current thread is 2 (Thread 5209.5216)]
 ...
 Completed event: 18900
 Current tick: 7979
 bt
 0x00007dc73a80cb88 in sqlite3OsWrite (id=0x2e1a75e8, pBuf=0x7dc72ebfe2dc, amt=12, offset=41552) at /usr/src/debug/sqlite-3.47.2-2.fc42.x86_64/sqlite3.c:26358
 26358     return id->pMethods->xWrite(id, pBuf, amt, offset);
 (rr) bt
 #0  0x00007dc73a80cb88 in sqlite3OsWrite (id=0x2e1a75e8, pBuf=0x7dc72ebfe2dc, amt=12, offset=41552) at /usr/src/debug/sqlite-3.47.2-2.fc42.x86_64/sqlite3.c:26358
 #1  syncJournal.part.0.lto_priv.0 (pPager=pPager@entry=0x2e1a7370, newHdr=newHdr@entry=0) at /usr/src/debug/sqlite-3.47.2-2.fc42.x86_64/sqlite3.c:61560
 #2  0x00007dc73a80cfaa in syncJournal (pPager=0x2e1a7370, newHdr=0) at /usr/src/debug/sqlite-3.47.2-2.fc42.x86_64/sqlite3.c:63629
 #3  sqlite3PagerCommitPhaseOne.part.0.lto_priv.0 (pPager=0x2e1a7370, zSuper=zSuper@entry=0x0, noSync=noSync@entry=0) at /usr/src/debug/sqlite-3.47.2-2.fc42.x86_64/sqlite3.c:63809
 #4  0x00007dc73a80d17f in sqlite3PagerCommitPhaseOne (pPager=<optimized out>, zSuper=0x0, noSync=0) at /usr/src/debug/sqlite-3.47.2-2.fc42.x86_64/sqlite3.c:63680
 #5  sqlite3BtreeCommitPhaseOne.part.0.lto_priv.0 (p=<optimized out>, zSuperJrnl=0x0) at /usr/src/debug/sqlite-3.47.2-2.fc42.x86_64/sqlite3.c:74958
 #6  0x00007dc73a70d5a0 in sqlite3BtreeCommitPhaseOne (p=<optimized out>, zSuperJrnl=0x0) at /usr/src/debug/sqlite-3.47.2-2.fc42.x86_64/sqlite3.c:74943
 #7  vdbeCommit (db=<optimized out>, p=0x7dc6f4006fb0) at /usr/src/debug/sqlite-3.47.2-2.fc42.x86_64/sqlite3.c:87982
 #8  sqlite3VdbeHalt (p=0x7dc6f4006fb0) at /usr/src/debug/sqlite-3.47.2-2.fc42.x86_64/sqlite3.c:88392
 #9  0x00007dc73a71d985 in sqlite3VdbeExec (p=p@entry=0x7dc6f4006fb0) at /usr/src/debug/sqlite-3.47.2-2.fc42.x86_64/sqlite3.c:94732
 #10 0x00007dc73a71eae4 in sqlite3Step (p=<optimized out>) at /usr/src/debug/sqlite-3.47.2-2.fc42.x86_64/sqlite3.c:91390
 #11 sqlite3_step (pStmt=<optimized out>) at /usr/src/debug/sqlite-3.47.2-2.fc42.x86_64/sqlite3.c:91451
 #12 sqlite3_step (pStmt=<optimized out>) at /usr/src/debug/sqlite-3.47.2-2.fc42.x86_64/sqlite3.c:91440
 #13 0x00000000004e8e71 in sqlite::Sql::Execute (this=0x7dc72ebfe760) at /usr/local/src/cvmfs/cvmfs/sql.cc:43
 #14 0x000000000044595f in catalog::WritableCatalog::UpdateNestedCatalog (this=0x2e1a6a70, path="/dir/inner_dir6", hash=..., size=57344, child_counters=...) at /usr/local/src/cvmfs/cvmfs/catalog_rw.cc:606
 #15 0x000000000042aae9 in catalog::WritableCatalogManager::SingleCatalogUploadCallback (this=0x7ffffb971930, result=...) at /usr/local/src/cvmfs/cvmfs/catalog_mgr_rw.cc:1674
 #16 0x0000000000439812 in BoundCallback<upload::SpoolerResult, catalog::WritableCatalogManager>::operator() (this=0x2e19da00, value=...) at /usr/local/src/cvmfs/cvmfs/util/async.h:91
 #17 0x000000000046ee8a in Observable<upload::SpoolerResult>::NotifyListeners (this=0x2e0a53e0, parameter=...) at /usr/local/src/cvmfs/cvmfs/util/concurrency_impl.h:174
 #18 0x00000000005c3c5b in upload::Spooler::ProcessingCallback (this=0x2e0a53e0, data=...) at /usr/local/src/cvmfs/cvmfs/upload.cc:133
 #19 0x00000000005c4436 in BoundCallback<upload::SpoolerResult, upload::Spooler>::operator() (this=0x2e0b8cb0, value=...) at /usr/local/src/cvmfs/cvmfs/util/async.h:91
 #20 0x000000000046ee8a in Observable<upload::SpoolerResult>::NotifyListeners (this=0x2e0bb5d0, parameter=...) at /usr/local/src/cvmfs/cvmfs/util/concurrency_impl.h:174
 #21 0x000000000046c1eb in IngestionPipeline::OnFileProcessed (this=0x2e0bb5d0, spooler_result=...) at /usr/local/src/cvmfs/cvmfs/ingestion/pipeline.cc:115
 #22 0x00000000004741c4 in BoundCallback<upload::SpoolerResult, IngestionPipeline>::operator() (this=0x2e0bc8c0, value=...) at /usr/local/src/cvmfs/cvmfs/util/async.h:91
 #23 0x000000000046ee8a in Observable<upload::SpoolerResult>::NotifyListeners (this=0x2e0badd0, parameter=...) at /usr/local/src/cvmfs/cvmfs/util/concurrency_impl.h:174
 #24 0x0000000000478c13 in TaskRegister::Process (this=0x2e0badc0, file_item=0x2e1cc850) at /usr/local/src/cvmfs/cvmfs/ingestion/task_register.cc:30
 #25 0x0000000000470c5d in TubeConsumer<FileItem>::MainConsumer (data=0x2e0badc0) at /usr/local/src/cvmfs/cvmfs/ingestion/task.h:56
 #26 0x00007dc739c23f54 in start_thread (arg=<optimized out>) at pthread_create.c:448
 #27 0x00007dc739ca7154 in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:100
 (rr)

 (rr) print $_siginfo
 $1 = {
   si_signo = 11,
   si_errno = 0,
   si_code = 1,
   _sifields = {
     ...
     _sigfault = {
       si_addr = 0x18
     },
   }
 }

 (rr) watch -l (void*) id->pMethods
 (rr) rc
 Continuing.
 [Switching to Thread 5209.5209]
 [Current thread is 1 (Thread 5209.5209)]
 #0  0x00007dc73a6e0af3 in memset (__dest=0x2e1a75e8, __ch=0, __len=120) at /usr/include/bits/string_fortified.h:59
 #1  closeUnixFile (id=0x2e1a75e8) at /usr/src/debug/sqlite-3.47.2-2.fc42.x86_64/sqlite3.c:40656
 #2  0x00007dc73a6e0cad in nolockClose (id=<optimized out>) at /usr/src/debug/sqlite-3.47.2-2.fc42.x86_64/sqlite3.c:40733
 Completed event: 18833
 Current tick: 6377324

 Thread 1 hit Hardware watchpoint 3: -location (void*) id->pMethods

 Old value = (void *) 0x0
 New value = (void *) 0x7dc73a8497c0 <nolockIoMethods>
 0x00007dc73a6e0af3 in memset (__dest=0x2e1a75e8, __ch=0, __len=120) at /usr/include/bits/string_fortified.h:59
 warning: Source file is more recent than executable.
 59        return __builtin___memset_chk (__dest, __ch, __len,
 (rr) bt
 #0  0x00007dc73a6e0af3 in memset (__dest=0x2e1a75e8, __ch=0, __len=120) at /usr/include/bits/string_fortified.h:59
 #1  closeUnixFile (id=0x2e1a75e8) at /usr/src/debug/sqlite-3.47.2-2.fc42.x86_64/sqlite3.c:40656
 #2  0x00007dc73a6e0cad in nolockClose (id=<optimized out>) at /usr/src/debug/sqlite-3.47.2-2.fc42.x86_64/sqlite3.c:40733
 #3  0x00007dc73a6eee5b in sqlite3OsClose (pId=0x2e1a75e8) at /usr/src/debug/sqlite-3.47.2-2.fc42.x86_64/sqlite3.c:26348
 #4  pager_end_transaction (pPager=0x2e1a7370, hasSuper=<optimized out>, bCommit=1) at /usr/src/debug/sqlite-3.47.2-2.fc42.x86_64/sqlite3.c:59323
 #5  0x00007dc73a6fc0a2 in sqlite3PagerCommitPhaseTwo (pPager=0x2e1a7370) at /usr/src/debug/sqlite-3.47.2-2.fc42.x86_64/sqlite3.c:63935
 #6  sqlite3BtreeCommitPhaseTwo (p=0x2e1a1ac0, bCleanup=<optimized out>) at /usr/src/debug/sqlite-3.47.2-2.fc42.x86_64/sqlite3.c:75044
 #7  0x00007dc73a70d5d3 in vdbeCommit (db=<optimized out>, p=0x2e1b2100) at /usr/src/debug/sqlite-3.47.2-2.fc42.x86_64/sqlite3.c:87994
 #8  sqlite3VdbeHalt (p=0x2e1b2100) at /usr/src/debug/sqlite-3.47.2-2.fc42.x86_64/sqlite3.c:88392
 #9  0x00007dc73a71af85 in sqlite3VdbeExec (p=p@entry=0x2e1b2100) at /usr/src/debug/sqlite-3.47.2-2.fc42.x86_64/sqlite3.c:97385
 #10 0x00007dc73a71eae4 in sqlite3Step (p=<optimized out>) at /usr/src/debug/sqlite-3.47.2-2.fc42.x86_64/sqlite3.c:91390
 #11 sqlite3_step (pStmt=<optimized out>) at /usr/src/debug/sqlite-3.47.2-2.fc42.x86_64/sqlite3.c:91451
 #12 sqlite3_step (pStmt=<optimized out>) at /usr/src/debug/sqlite-3.47.2-2.fc42.x86_64/sqlite3.c:91440
 #13 0x00000000004e8e71 in sqlite::Sql::Execute (this=0x2e1b5ef0) at /usr/local/src/cvmfs/cvmfs/sql.cc:43
 #14 0x00000000004418ac in sqlite::Database<catalog::CatalogDatabase>::CommitTransaction (this=0x2e19b2b0) at /usr/local/src/cvmfs/cvmfs/sql_impl.h:275
 #15 0x00000000004433d0 in catalog::WritableCatalog::Commit (this=0x2e1a6a70) at /usr/local/src/cvmfs/cvmfs/catalog_rw.cc:83
 #16 0x0000000000428e8e in catalog::WritableCatalogManager::FinalizeCatalog (this=0x7ffffb971930, catalog=0x2e1a6a70, stop_for_tweaks=false) at /usr/local/src/cvmfs/cvmfs/catalog_mgr_rw.cc:1284
 #17 0x000000000042a6f4 in catalog::WritableCatalogManager::ScheduleReadyCatalogs (this=0x7ffffb971930) at /usr/local/src/cvmfs/cvmfs/catalog_mgr_rw.cc:1632
 #18 0x000000000052eeb0 in swissknife::IngestSQL::do_additions (this=0x2e09f7a0, all_dirs=std::map with 12 elements = {...}, all_files=std::map with 9 elements = {...}, all_symlinks=std::map with 0 elements,
     lease_path="", catalog_manager=...) at /usr/local/src/cvmfs/cvmfs/swissknife_ingestsql.cc:1133
 #19 0x000000000052e6ad in swissknife::IngestSQL::process_sqlite (this=0x2e09f7a0, dbs=std::vector of length 1, capacity 1 = {...}, catalog_manager=..., allow_additions=true, allow_deletions=true, lease_path="",
     additional_prefix="") at /usr/local/src/cvmfs/cvmfs/swissknife_ingestsql.cc:1038
 #20 0x000000000052db9a in swissknife::IngestSQL::Main (this=0x2e09f7a0, args=std::map with 12 elements = {...}) at /usr/local/src/cvmfs/cvmfs/swissknife_ingestsql.cc:889
 #21 0x00000000005540e6 in main (argc=24, argv=0x7ffffb9731b8) at /usr/local/src/cvmfs/cvmfs/swissknife_main.cc:190

andrey-utkin committed:
catalog_wr: Fix recursive locking caused by previous change

Now that WritableCatalog::Transaction() locks, calling it from
WritableCatalog::UpdateNestedCatalog() hangs, because this is not a
recursive mutex so obtaining the mutex for the second time doesn't work.

 Thread 6 (Thread 2776.2783 (cvmfs_swissknife_debug)):
 #0  0x0000000070000002 in syscall_traced ()
 #1  0x000077c41c140fd8 in _raw_syscall () at /builddir/build/BUILD/rr-5.9.0-build/rr-5.9.0/src/preload/raw_syscall.S:120
 #2  0x000077c41c13b2fd in traced_raw_syscall (call=<optimized out>) at /builddir/build/BUILD/rr-5.9.0-build/rr-5.9.0/src/preload/syscallbuf.c:379
 #3  0x000077c41c14020b in syscall_hook_internal (call=0x77c4183fffa0) at /builddir/build/BUILD/rr-5.9.0-build/rr-5.9.0/src/preload/syscallbuf.c:4203
 #4  syscall_hook (call=0x77c4183fffa0) at /builddir/build/BUILD/rr-5.9.0-build/rr-5.9.0/src/preload/syscallbuf.c:4387
 #5  syscall_hook (call=0x77c4183fffa0) at /builddir/build/BUILD/rr-5.9.0-build/rr-5.9.0/src/preload/syscallbuf.c:4371
 #6  0x000077c41c13a5c3 in _syscall_hook_trampoline () at /builddir/build/BUILD/rr-5.9.0-build/rr-5.9.0/src/preload/syscall_hook.S:308
 #7  0x000077c41c13a62d in __morestack () at /builddir/build/BUILD/rr-5.9.0-build/rr-5.9.0/src/preload/syscall_hook.S:443
 #8  0x000077c41c13a649 in _syscall_hook_trampoline_48_3d_00_f0_ff_ff () at /builddir/build/BUILD/rr-5.9.0-build/rr-5.9.0/src/preload/syscall_hook.S:462
 #9  0x000077c41b311006 in __lll_lock_wait () from /lib64/libc.so.6
 #10 0x000077c41b317501 in pthread_mutex_lock@@GLIBC_2.2.5 () from /lib64/libc.so.6
 #11 0x000000000040b8ef in RAII<pthread_mutex_t, (RAII_Polymorphism::T)0>::Enter (this=0x77c407ffe618) at /usr/local/src/cvmfs/cvmfs/util/mutex.h:77
 #12 0x000000000040e74f in RAII<pthread_mutex_t, (RAII_Polymorphism::T)0>::RAII (this=0x77c407ffe618, object=0x2e05ed80) at /usr/local/src/cvmfs/cvmfs/util/mutex.h:45
 #13 0x0000000000443334 in catalog::WritableCatalog::Transaction (this=0x2e062560) at /usr/local/src/cvmfs/cvmfs/catalog_rw.cc:75
 #14 0x000000000042bf54 in catalog::WritableCatalog::SetDirty (this=0x2e062560) at /usr/local/src/cvmfs/cvmfs/catalog_rw.h:164
 #15 0x00000000004458bb in catalog::WritableCatalog::UpdateNestedCatalog (this=0x2e062560, path="/dir/inner_dir6", hash=..., size=57344, child_counters=...) at /usr/local/src/cvmfs/cvmfs/catalog_rw.cc:598
 #16 0x000000000042aae9 in catalog::WritableCatalogManager::SingleCatalogUploadCallback (this=0x7ffdefbe23a0, result=...) at /usr/local/src/cvmfs/cvmfs/catalog_mgr_rw.cc:1674
 #17 0x0000000000439812 in BoundCallback<upload::SpoolerResult, catalog::WritableCatalogManager>::operator() (this=0x2dfc5950, value=...) at /usr/local/src/cvmfs/cvmfs/util/async.h:91
 #18 0x000000000046eede in Observable<upload::SpoolerResult>::NotifyListeners (this=0x2decd3f0, parameter=...) at /usr/local/src/cvmfs/cvmfs/util/concurrency_impl.h:174
 #19 0x00000000005c3caf in upload::Spooler::ProcessingCallback (this=0x2decd3f0, data=...) at /usr/local/src/cvmfs/cvmfs/upload.cc:133
 #20 0x00000000005c448a in BoundCallback<upload::SpoolerResult, upload::Spooler>::operator() (this=0x2dee00b0, value=...) at /usr/local/src/cvmfs/cvmfs/util/async.h:91
 #21 0x000000000046eede in Observable<upload::SpoolerResult>::NotifyListeners (this=0x2dee2980, parameter=...) at /usr/local/src/cvmfs/cvmfs/util/concurrency_impl.h:174
 #22 0x000000000046c23f in IngestionPipeline::OnFileProcessed (this=0x2dee2980, spooler_result=...) at /usr/local/src/cvmfs/cvmfs/ingestion/pipeline.cc:115
 #23 0x0000000000474218 in BoundCallback<upload::SpoolerResult, IngestionPipeline>::operator() (this=0x2dee1d20, value=...) at /usr/local/src/cvmfs/cvmfs/util/async.h:91
 #24 0x000000000046eede in Observable<upload::SpoolerResult>::NotifyListeners (this=0x2dee4570, parameter=...) at /usr/local/src/cvmfs/cvmfs/util/concurrency_impl.h:174
 #25 0x0000000000478c67 in TaskRegister::Process (this=0x2dee4560, file_item=0x2e048740) at /usr/local/src/cvmfs/cvmfs/ingestion/task_register.cc:30
 #26 0x0000000000470cb1 in TubeConsumer<FileItem>::MainConsumer (data=0x2dee4560) at /usr/local/src/cvmfs/cvmfs/ingestion/task.h:56
 #27 0x000077c41b313f54 in start_thread () from /lib64/libc.so.6
 #28 0x000077c41b397154 in clone () from /lib64/libc.so.6
 (rr) thread 6
 [Switching to thread 6 (Thread 2776.2783)]
 #0  0x0000000070000002 in syscall_traced ()
 (rr) frame 11
 #11 0x000000000040b8ef in RAII<pthread_mutex_t, (RAII_Polymorphism::T)0>::Enter (this=0x77c407ffe618) at /usr/local/src/cvmfs/cvmfs/util/mutex.h:77
 77        pthread_mutex_lock(&ref_);
 (rr) print ref_->__data.__owner
 $3 = 2783

andrey-utkin committed:
catalog_wr: Fix recursive locking caused by previous change

Now that WritableCatalog::Transaction() locks, calling it from
WritableCatalog::UpdateNestedCatalog() hangs, because this is not a
recursive mutex so obtaining the mutex for the second time doesn't work.

 Thread 6 (Thread 2776.2783 (cvmfs_swissknife_debug)):
 #0  0x0000000070000002 in syscall_traced ()
 #1  0x000077c41c140fd8 in _raw_syscall () at /builddir/build/BUILD/rr-5.9.0-build/rr-5.9.0/src/preload/raw_syscall.S:120
 #2  0x000077c41c13b2fd in traced_raw_syscall (call=<optimized out>) at /builddir/build/BUILD/rr-5.9.0-build/rr-5.9.0/src/preload/syscallbuf.c:379
 #3  0x000077c41c14020b in syscall_hook_internal (call=0x77c4183fffa0) at /builddir/build/BUILD/rr-5.9.0-build/rr-5.9.0/src/preload/syscallbuf.c:4203
 #4  syscall_hook (call=0x77c4183fffa0) at /builddir/build/BUILD/rr-5.9.0-build/rr-5.9.0/src/preload/syscallbuf.c:4387
 #5  syscall_hook (call=0x77c4183fffa0) at /builddir/build/BUILD/rr-5.9.0-build/rr-5.9.0/src/preload/syscallbuf.c:4371
 #6  0x000077c41c13a5c3 in _syscall_hook_trampoline () at /builddir/build/BUILD/rr-5.9.0-build/rr-5.9.0/src/preload/syscall_hook.S:308
 #7  0x000077c41c13a62d in __morestack () at /builddir/build/BUILD/rr-5.9.0-build/rr-5.9.0/src/preload/syscall_hook.S:443
 #8  0x000077c41c13a649 in _syscall_hook_trampoline_48_3d_00_f0_ff_ff () at /builddir/build/BUILD/rr-5.9.0-build/rr-5.9.0/src/preload/syscall_hook.S:462
 #9  0x000077c41b311006 in __lll_lock_wait () from /lib64/libc.so.6
 #10 0x000077c41b317501 in pthread_mutex_lock@@GLIBC_2.2.5 () from /lib64/libc.so.6
 #11 0x000000000040b8ef in RAII<pthread_mutex_t, (RAII_Polymorphism::T)0>::Enter (this=0x77c407ffe618) at /usr/local/src/cvmfs/cvmfs/util/mutex.h:77
 #12 0x000000000040e74f in RAII<pthread_mutex_t, (RAII_Polymorphism::T)0>::RAII (this=0x77c407ffe618, object=0x2e05ed80) at /usr/local/src/cvmfs/cvmfs/util/mutex.h:45
 #13 0x0000000000443334 in catalog::WritableCatalog::Transaction (this=0x2e062560) at /usr/local/src/cvmfs/cvmfs/catalog_rw.cc:75
 #14 0x000000000042bf54 in catalog::WritableCatalog::SetDirty (this=0x2e062560) at /usr/local/src/cvmfs/cvmfs/catalog_rw.h:164
 #15 0x00000000004458bb in catalog::WritableCatalog::UpdateNestedCatalog (this=0x2e062560, path="/dir/inner_dir6", hash=..., size=57344, child_counters=...) at /usr/local/src/cvmfs/cvmfs/catalog_rw.cc:598
 #16 0x000000000042aae9 in catalog::WritableCatalogManager::SingleCatalogUploadCallback (this=0x7ffdefbe23a0, result=...) at /usr/local/src/cvmfs/cvmfs/catalog_mgr_rw.cc:1674
 #17 0x0000000000439812 in BoundCallback<upload::SpoolerResult, catalog::WritableCatalogManager>::operator() (this=0x2dfc5950, value=...) at /usr/local/src/cvmfs/cvmfs/util/async.h:91
 #18 0x000000000046eede in Observable<upload::SpoolerResult>::NotifyListeners (this=0x2decd3f0, parameter=...) at /usr/local/src/cvmfs/cvmfs/util/concurrency_impl.h:174
 #19 0x00000000005c3caf in upload::Spooler::ProcessingCallback (this=0x2decd3f0, data=...) at /usr/local/src/cvmfs/cvmfs/upload.cc:133
 #20 0x00000000005c448a in BoundCallback<upload::SpoolerResult, upload::Spooler>::operator() (this=0x2dee00b0, value=...) at /usr/local/src/cvmfs/cvmfs/util/async.h:91
 #21 0x000000000046eede in Observable<upload::SpoolerResult>::NotifyListeners (this=0x2dee2980, parameter=...) at /usr/local/src/cvmfs/cvmfs/util/concurrency_impl.h:174
 #22 0x000000000046c23f in IngestionPipeline::OnFileProcessed (this=0x2dee2980, spooler_result=...) at /usr/local/src/cvmfs/cvmfs/ingestion/pipeline.cc:115
 #23 0x0000000000474218 in BoundCallback<upload::SpoolerResult, IngestionPipeline>::operator() (this=0x2dee1d20, value=...) at /usr/local/src/cvmfs/cvmfs/util/async.h:91
 #24 0x000000000046eede in Observable<upload::SpoolerResult>::NotifyListeners (this=0x2dee4570, parameter=...) at /usr/local/src/cvmfs/cvmfs/util/concurrency_impl.h:174
 #25 0x0000000000478c67 in TaskRegister::Process (this=0x2dee4560, file_item=0x2e048740) at /usr/local/src/cvmfs/cvmfs/ingestion/task_register.cc:30
 #26 0x0000000000470cb1 in TubeConsumer<FileItem>::MainConsumer (data=0x2dee4560) at /usr/local/src/cvmfs/cvmfs/ingestion/task.h:56
 #27 0x000077c41b313f54 in start_thread () from /lib64/libc.so.6
 #28 0x000077c41b397154 in clone () from /lib64/libc.so.6
 (rr) thread 6
 [Switching to thread 6 (Thread 2776.2783)]
 #0  0x0000000070000002 in syscall_traced ()
 (rr) frame 11
 #11 0x000000000040b8ef in RAII<pthread_mutex_t, (RAII_Polymorphism::T)0>::Enter (this=0x77c407ffe618) at /usr/local/src/cvmfs/cvmfs/util/mutex.h:77
 77        pthread_mutex_lock(&ref_);
 (rr) print ref_->__data.__owner
 $3 = 2783

christge-sft commented on cvmfs/catalog_downloader.cc:
```diff
@@ -5,11 +5,11 @@ int kCatalogDownloadMultiplier = 32;
+  shash::Any catalog_hash;
   // will PANIC if download failed
-  // catalog_mgr_->LoadCatalog(PathString("") /* not used */, *input->GetHash(),
-  // &catalog_path, &catalog_hash, NULL);
```
I understand that this was commented out because download failure was not behaving properly. Now that the code is back, is it tested against this scenario?

andrey-utkin replied:
No such test AFAIK.
This comment refers to the fact that `SimpleCatalogManager::LoadCatalog()` using PANIC() for every unsuccessful condition. I am going to add graceful runtime error handling there.

andrey-utkin replied:
I've replaced `PANIC()`s with returning an error in `SimpleCatalogManager::LoadCatalog()`.
[It's unclear how to raise an error from `TaskCatalogDownload::Process()` as it returns void and it's an inherited method](https://github.com/cvmfs/cvmfs/issues/4048).
Apparently it's not even necessary to check return value of it in the caller, `TaskCatalogDownload::Process()`, because it then calls `WritableCatalogManager::CatalogDownloadCallback()` which does `catalog_download_map_.find(result.hash)` and finds out. The problem is that it just does an assert() on it so basically PANIC again :) I've now replaced it with `return;` as below.
I will work on writing a test for this.

christge-sft commented on cvmfs/ingestion/pipeline.cc:
```diff
@@ -258,3 +258,117 @@ void ScrubbingPipeline::Spawn() {
+{
+  // we don't need much parallelism for this pipeline
+  const unsigned int nfork_base = 1;
+
```
What each of these loops is supposed to be doing? Also what's behind the `nfork_base * <coef>`?

andrey-utkin replied:
This allocates the workers for a segmented pipeline.
The `nfork_base * <coef>` determines how many workers are created, `coef` represents either an empyrical or hunch-based optimal value. I will consult with the authors for more.
Probably makes sense for very high load or big batch jobs.

andrey-utkin replied:
> This allocates the workers for a segmented pipeline. The `nfork_base * <coef>` determines how many workers are created, `coef` represents either an empyrical or hunch-based optimal value. I will consult with the authors for more. Probably makes sense for very high load or big batch jobs.

So Matt Harvey confirmed this. All the coefficients are ad-hoc picked, and we are free to change them. The only hard fact is that we in principle can allocate the workers in any quantity we want.

christge-sft commented on cvmfs/swissknife_ingestsql.cc:
```diff
@@ -648,7 +648,11 @@ int swissknife::IngestSQL::Main(const swissknife::ArgumentList &args) {
+    if (config_map.find("CVMFS_HTTP_PROXY") != config_map.end()) {
+      proxy = retrieve_config(config_map, "CVMFS_HTTP_PROXY");
+    } else {
+      proxy = "DIRECT";
```
I think that it's a deliberate decision to discourage users working on `DIRECT` mode, and somehow "force" them to set up their proxies in order to relieve Stratum1's from extreme workloads. Do we want to loose on this policy here?

andrey-utkin replied:
If this is the maintainers' policy, I am ready to follow it and change this. @vvolkl please confirm.

christge-sft commented on cvmfs/swissknife_ingestsql.cc:
```diff
@@ -441,9 +442,12 @@ static XattrList marshal_xattrs(const char *acl_string) {
+      LogCvmfs(kLogCvmfs, kLogStderr,
+               "failed to set system.posix_acl_access (ACL size %ld)",
+               binary_size);
+    }
```
In the new version, execution won't abort after logging. And in this case we can't change the return value of the function to indicate failure.

andrey-utkin replied:
My fault, I will add error handling, changing the function's signature.

christge-sft commented on cvmfs/swissknife_ingestsql.cc:
```diff
@@ -517,6 +521,7 @@ static vector<string> get_file_list(string &path) {
 extern bool g_log_with_time;
 
 int swissknife::IngestSQL::Main(const swissknife::ArgumentList &args) {
+  int ret;
```
where may we find a return value interpretation?

andrey-utkin replied:
Here it's a variable to store the return value of called functions.
I can move its declaration closer to its use, this just tends to cause annoyance later sometimes, when similar value storing and using is needed earlier in the function, and ...::Main is a long one.

christge-sft commented:
I tried to install the tools as described inside the `README.md` and installation failed. (I figured this out since all tests were failing individually). 

I tried both to install them all together with `sudo go install ./...` in the parent repo as well as one by one. Has anyone else faced the same problem? The installation script didn't issue any warning.
```
$ sudo go install ./cmd/cvmfs_chmod ; echo $? ; whereis cvmfs_chmod
0
cvmfs_chmod:
```

andrey-utkin committed:
test/src/902-cvmfs-posix-tools: fix check for varnish presence

andrey-utkin committed:
test/src/902-cvmfs-posix-tools: fix check for varnish presence

andrey-utkin committed:
ingestsql: fix log message on giving up waiting

It was saying:
"Out of time waiting for the mount to reach revision <current revision>"

Now it says:
"Out of time waiting for the mount to reach revision <waited upon revision>"

Fixes: 184a2ecb453e ("ingestsql: add option -W for wait timeout")

andrey-utkin commented:
Huge thanks for the review, @christge-sft !


> I tried to install the tools as described inside the `README.md` and installation failed. (I figured this out since all tests were failing individually).

I will check README.md and correct as needed.
The way to build and install these tools is to add `-DBUILD_POSIX_TOOLS` to cmake options.

andrey-utkin committed:
cvmfs-posix-tools: refine documentation

andrey-utkin committed:
cvmfs-posix-tools: refine documentation

andrey-utkin committed:
ingestsql: handle errors in marshal_xattrs gracefully

andrey-utkin committed:
ingestsql: handle errors in marshal_xattrs gracefully

DrDaveD commented:
> perhaps it's OK to have a completely different tool by the same name if they're only installed on different machines.  rpm won't let both packages be installed on the same machine.

Note that stratum 0 machines install both the cvmfs client and cvmfs-server, so these would need to be packaged in a separate package and not be in the same package with the cvmfs client.

andrey-utkin committed:
test: t_acl: fix build warning

In file included from /usr/local/src/cvmfs/_deps/googletest-src/googletest/include/gtest/gtest.h:71,
		 from /usr/local/src/cvmfs/test/unittests/t_acl.cc:9:
/usr/local/src/cvmfs/test/unittests/t_acl.cc: In function 'void should_pass_noncompat(const char*, unsigned char*, unsigned int)':
/usr/local/src/cvmfs/test/unittests/t_acl.cc:34:24: warning: argument 2 null where non-null expected [-Wnonnull]
   34 |     ASSERT_EQ(0, memcmp(binary_acl, acl_binary_expected, binary_size));
      |                  ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/local/src/cvmfs/_deps/googletest-src/googletest/include/gtest/internal/gtest-port.h:300,
		 from /usr/local/src/cvmfs/_deps/googletest-src/googletest/include/gtest/gtest-message.h:57,
		 from /usr/local/src/cvmfs/_deps/googletest-src/googletest/include/gtest/gtest-assertion-result.h:46,
		 from /usr/local/src/cvmfs/_deps/googletest-src/googletest/include/gtest/gtest.h:63:
/usr/include/string.h:64:12: note: in a call to function 'int memcmp(const void*, const void*, size_t)' declared 'nonnull'
   64 | extern int memcmp (const void *__s1, const void *__s2, size_t __n)

andrey-utkin committed:
test: t_acl: fix build warning

In file included from /usr/local/src/cvmfs/_deps/googletest-src/googletest/include/gtest/gtest.h:71,
		 from /usr/local/src/cvmfs/test/unittests/t_acl.cc:9:
/usr/local/src/cvmfs/test/unittests/t_acl.cc: In function 'void should_pass_noncompat(const char*, unsigned char*, unsigned int)':
/usr/local/src/cvmfs/test/unittests/t_acl.cc:34:24: warning: argument 2 null where non-null expected [-Wnonnull]
   34 |     ASSERT_EQ(0, memcmp(binary_acl, acl_binary_expected, binary_size));
      |                  ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/local/src/cvmfs/_deps/googletest-src/googletest/include/gtest/internal/gtest-port.h:300,
		 from /usr/local/src/cvmfs/_deps/googletest-src/googletest/include/gtest/gtest-message.h:57,
		 from /usr/local/src/cvmfs/_deps/googletest-src/googletest/include/gtest/gtest-assertion-result.h:46,
		 from /usr/local/src/cvmfs/_deps/googletest-src/googletest/include/gtest/gtest.h:63:
/usr/include/string.h:64:12: note: in a call to function 'int memcmp(const void*, const void*, size_t)' declared 'nonnull'
   64 | extern int memcmp (const void *__s1, const void *__s2, size_t __n)

andrey-utkin committed:
catalog_mgr_ro: handle errors gracefully

This commit removes PANIC() instances introduced earlier.

andrey-utkin committed:
catalog_mgr_ro: handle errors gracefully

This commit removes PANIC() instances introduced earlier.

andrey-utkin committed:
catalog_mgr_rw: Handle download error gracefully in CatalogDownloadCallback()

andrey-utkin committed:
catalog_mgr_rw: Handle download error gracefully in CatalogDownloadCallback()

andrey-utkin commented:
Hi @christge-sft , I think I have addressed the questions and change requests, could you please give another round of review?

andrey-utkin committed:
cvmfs-posix-tools: test: make it compatible with fstab entry format starting with "cvmfs2#"

Test was developed against upstream with this change merged:
https://github.com/cvmfs/cvmfs/pull/3903 (cvmfs_server: fstab: change
type from fuse to cvmfs, abandon cvmfs2# prefix in name)

andrey-utkin committed:
cvmfs-posix-tools: test: make it compatible with fstab entry format starting with "cvmfs2#"

Test was developed against upstream with this change merged:
https://github.com/cvmfs/cvmfs/pull/3903 (cvmfs_server: fstab: change
type from fuse to cvmfs, abandon cvmfs2# prefix in name)

andrey-utkin committed:
cvmfs-posix-tools: test: pass through all command invocations' stdout & stderr

andrey-utkin committed:
cvmfs-posix-tools: test: adapt to non-root running

These are the tweaks which it took to run the test successfully in the
environment similar to Github Actions CI.

Previously these tests were exercised only under root user.

Possibly some sudoing is excessive, but let's get it to work first.

andrey-utkin committed:
test: 902-cvmfs-posix-tools: use curl options supported in Almalinux 8 version

andrey-utkin committed:
cvmfs-posix-tools: test: work around lingering cvmfs2 on Debian-like

andrey-utkin committed:
cvmfs-posix-tools: test: drop unused config files

andrey-utkin committed:
test: s3: enable src/9*

andrey-utkin committed:
cvmfs-posix-tools: test: support ci_s3_integration workflow

Proxied mode disabled for now

andrey-utkin committed:
cvmfs-posix-tools: include all source files into cmake dependencies

Previously, changes in e.g. pkg/*.go wouldn't trigger a rebuild.

andrey-utkin committed:
Change RPM packaging and Dockerfile-dev for cvmfs-posix-tools

RPM changes cause the established shell script cvmfs_rsync to be not
available anymore, so this is probably not finished matter and some
package splitting or some other way of disambiguation will have to be
applied. This change focuses on getting cvmfs-posix-tools to build,
install and run the test in ci_s3_integration github actions workflow.

andrey-utkin committed:
test/src/902-cvmfs-posix-tools: add quick tag
